internal/controller/garbagecollector_controller.go

```diff
@@ -0,0 +1,141 @@
+/*
+Copyright 2024.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+	"time"
+
+	velerov1 "github.com/vmware-tanzu/velero/pkg/apis/velero/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/utils/ptr"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	nacv1alpha1 "github.com/migtools/oadp-non-admin/api/v1alpha1"
+	"github.com/migtools/oadp-non-admin/internal/common/constant"
+	"github.com/migtools/oadp-non-admin/internal/common/function"
+	"github.com/migtools/oadp-non-admin/internal/predicate"
+)
+
+// GarbageCollectorReconciler reconciles Velero objects
+type GarbageCollectorReconciler struct {
+	client.Client
+	Scheme        *runtime.Scheme
+	OADPNamespace string
+	Frequency     time.Duration
+}
+
+var previousGarbageCollectorRun *time.Time
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+func (r *GarbageCollectorReconciler) Reconcile(ctx context.Context, _ ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+
+	labelSelector := client.MatchingLabels{
+		constant.OadpLabel:      constant.OadpLabelValue,
+		constant.ManagedByLabel: constant.ManagedByLabelValue,
+	}
+
+	if previousGarbageCollectorRun == nil || time.Now().After(previousGarbageCollectorRun.Add(r.Frequency)) {
+		previousGarbageCollectorRun = ptr.To(time.Now())
+		logger.V(1).Info("Garbage Collector Reconcile start")
+		// TODO duplication in delete logic
+		// TODO do deletion in parallel?
+
+		// TODO delete BSL
+		// TODO delete secret as well?
+
+		veleroBackupList := &velerov1.BackupList{}
+		if err := r.List(ctx, veleroBackupList, client.InNamespace(r.OADPNamespace), labelSelector); err != nil {
+			logger.Error(err, "Unable to fetch Backups in OADP namespace")
+			return ctrl.Result{}, err
+		}
+		for _, backup := range veleroBackupList.Items {
+			annotations := backup.GetAnnotations()
+			// TODO check UUID as well?
+			if !function.CheckVeleroBackupAnnotations(&backup) {
+				logger.V(1).Info("Backup does not have required annotations", constant.NameString, backup.Name)
+				// TODO delete?
```

Comment:
object has NAC labels, but no NAC annotations. What should we do?

Reply:
This should not happen https://github.com/migtools/oadp-non-admin/blob/25e40d6abd8c49c43a1a89a9f49d6d489d07707e/internal/controller/nonadminbackup_controller.go#L638 but if for some reason it does I think it would be logged as warning or INFO not a DEBUG. Alternative if you want this to be visible in OpenShift’s Events tab, you can emit an Event from your controller. This will help cluster admins and users see what’s happening without digging into logs.

Reply:
I think there is no warning level in this lib we use

what is the decision here?

Reply:
Annotations are only to ensure ownership reference exists between NA* and Velero objects.
If labels are there it means that they are managed and created by the controller.
If annotations are missing it means that for some different update/patch to the Velero object the ownership reference was removed.

On a second thought I believe this is something that is going to be subject for GC removal.

Even if the Velero Syncs back objects from the S3 storage to the openshift-adp namespace, those annotations are preserved.